lib/pbench/server/database/models/users.py

```diff
@@ -119,7 +119,10 @@ def delete(username):
         """
         try:
             user_query = Database.db_session.query(User).filter_by(username=username)
-            user_query.delete()
+            rows_deleted = user_query.delete()
+            if rows_deleted == 0:
+                return False
+
```

Comment:
@webbnh the `.delete()` wont immediately delete the record, to delete we have to `commit` the operation (which we do on line 126). We only commit if we found the user entry to delete and then return true (meaning there was a user present in table and we deleted it). The `.delete()` method actually is a kind of existence check before trying the deletion, it wont delete any record unless we `commit` it. If we didn't find any existing record the function simply return false without committing any changes as if nothing happened.  

`return true` means the changes are committed and the user is no longer there and `return false` means we couldn't find the user because it no longer exist in the db. If two people happens to perform the delete operation on the same user at the exact same time, one of them is bound to fail as the second operation might be a fraction late, but isn't that desirable outcome?

Reply:
If Alice wants to delete a user from the database, and if, while she's in the process of performing the operation, Bob deletes the user, then Alice's aims have been accomplished, so why should the tool report a failure?

On the other hand, if Alice wants to delete a user from the database, but she mistypes the username such that it doesn't match any user, then the tool _should_ report a failure, so that she knows that her aims have _not_ been accomplished.

There are two separate layers in play here:  there is the tool (`pbench-user-delete`) and there is the pbench library code which it calls.  In your proposed code, there is no way for the tool to tell the difference between a user which never existed and a user which was deleted by someone else.

Perhaps the way out of my conundrum is to check between the query (line 121) and the deletion (line 122) that we actually have a user that we're trying to delete.  (And, it might not be silly to have a sanity check that we're deleting exactly one!)  If the query comes back empty, we can return an indication that the target user does not exist (perhaps by raising an exception).  The tool can catch this and report an error to the user.  If the query looks good but after that the deletion fails (because the target user no longer exists), or if the commit fails, or something else goes wrong, then we can try to address the situation appropriately, but it won't represent a user error, and it won't necessarily indicate a tool failure (it depends on the exact nature of the problem).

Reply:
> If Alice wants to delete a user from the database, and if, while she's in the process of performing the operation, Bob deletes the user, then Alice's aims have been accomplished, so why should the tool report a failure?

> On the other hand, if Alice wants to delete a user from the database, but she mistypes the username such that it doesn't match any user, then the tool should report a failure, so that she knows that her aims have not been accomplished.

I believe this is an extreme edge case where this kind of race condition is happening. I am also fairly certain that if Alice is in the middle of the delete operation (meaning she hit the command first) she will get the success. However, Bob will get an error since at this time there is no such a user and username would be considered as a mistyped one. 

> There are two separate layers in play here: there is the tool (`pbench-user-delete`) and there is the pbench library code which it calls. In your proposed code, there is no way for the tool to tell the difference between a user which never existed and a user which was deleted by someone else.

We unfortunately can't differentiate between user never existed and a user which was deleted by someone else. As soon as someone deletes the user, that user becomes _never existed_ one right? 

> Perhaps the way out of my conundrum is to check between the query (line 121) and the deletion (line 122) that we actually have a user that we're trying to delete. (And, it might not be silly to have a sanity check that we're deleting exactly one!) If the query comes back empty, we can return an indication that the target user does not exist (perhaps by raising an exception). The tool can catch this and report an error to the user. If the query looks good but after that the deletion fails (because the target user no longer exists), or if the commit fails, or something else goes wrong, then we can try to address the situation appropriately, but it won't represent a user error, and it won't necessarily indicate a tool failure (it depends on the exact nature of the problem).

Are you suggesting we return an exception instead of boolean `False` if we dont have a user entry to delete? Because we can check the delete operation return value for `False` condition and report an error to a user. The reason I chose the boolean value to return instead of raising an exception is that we would have to make a changes in `users_api.py` file. Right now with `delete` API call we dont report the end user that the `username doesn't exist` if username provided does not match our db record. We just return 200 since operationally we did not failed (I think we decided this because we didn't want someone to trial and error on our database via delete API find find any valid username).

Reply:
Yes, I do think that "user doesn't exist" should be reported, whether via CLI or API. You're right that it might just be a race condition where someone else deleted it, but that's fairly unlikely in real use; a typo is more likely, and letting the user walk away thinking they might have successfully deleted the wrong username probably isn't the best user experience.

Reply:
I agree that I'm quite possibly splitting hairs, here.

> I am also fairly certain that if Alice is in the middle of the delete operation (meaning she hit the command first) she will get the success.

I agree that there might be several different possibilities for how the timing could work out.  And, among them, Alice could see a failure or Bob could see a failure.  (And, it doesn't really matter which of them sees the failure, since they are both users and we want the experience to be best for each of them.)  However, by making some subtle choices in the code, I think we can allow one of the timing sequences to display success for _both_ of them, despite the fact that there was only one deletion between them.

What I'm suggesting is to have the code check for existence before the deletion, and base the result on that, rather than basing the result on whether the deletion itself was successful.  That is, if the username doesn't exist before the deletion, that is a user error which needs to be reported; however, if the username exists, then it doesn't matter (to the end-user) whether the delete operation itself succeeded or not, so long as the username is no longer in the database when the shell prompt returns.  So, if the query returns one row and the deletion deletes none, then I would consider that to be a "success" case, because the row is now gone.

(Disclaimer:  without digging deep into SQLAlchemy, I have no understanding of how it is implemented, what the possibilities for concurrent access are, how it administers transactions and ensures atomicity, etc., so this might all be moot...and, if it is, I apologize for the kerfuffle; but I think that a "failure to delete" yields an ambiguous situation for the user, while "does not exist" is a state which is clear, so I'm trying to steer us toward the latter.)

Reply:
Okay this is about the longest single comment that I have seen for 3 lines of code (4 if you count the extra blank line being added).

We seem to be mixing how we implement the delete mechanism from the interface we offer in the CLI (which begs the question, why is the CLI not making API calls?).

We can all probably agree that maintaining the principle of _separation of concerns_ is worth following here.  As such, we'd want to keep the concern of how we present results of operations performed to the users separate from the concern of how we implement those operations.

We use a generic `Exception` at line 128 to roll back the operation and re-raise the exception.  What happens when `commit()` fails to commit the transaction of deleting the user?  Does it raise and exception?

Perhaps we can do something instead like the following:

```
        user_query = Database.db_session.query(User).filter_by(username=username)
        if not user_query:
            return False

        try:
            user_query.delete()
            Database.db_session.commit()
        except ExceptionRaisedWhenTransctionIsAlreadyCompleted:
            Database.db_session.rollback()

        return True
```

Reply:
> why is the CLI not making API calls?

Because...the API requires authentication, whereas elements of this CLI are required to run before authentication is available (i.e., this is how we create the primordial users); this CLI only runs on the same node as the API, so it doesn't need network access (and doesn't need to be authenticated, under the presumption that access to the system is restricted); and, this CLI _can_ directly access the database, so it doesn't need to go through the API.

I like your proposed code, except that, rather than return `False`, I would have it raise an exception, since I would regard that as an "exceptional" case.

Reply:
Addressed Peter's suggestion